pkg/controller/atlasdatafederation/datafederation_controller.go

```diff
@@ -0,0 +1,273 @@
+				return result.ReconcileResult(), nil
+			}
+		} else {
+			return result.ReconcileResult(), nil
+		}
+	}
+
+	ctx.SetConditionTrue(status.ReadyType)
+	return workflow.OK().ReconcileResult(), nil
+}
+
+func (r *AtlasDataFederationReconciler) removeDeletionFinalizer(context context.Context, df *mdbv1.AtlasDataFederation) error {
+	err := r.Client.Get(context, kube.ObjectKeyFromObject(df), df)
+	if err != nil {
+		return fmt.Errorf("cannot get AtlasDeployment while adding finalizer: %w", err)
+	}
+
+	customresource.UnsetFinalizer(df, customresource.FinalizerLabel)
+	if err = r.Client.Update(context, df); err != nil {
+		return fmt.Errorf("failed to remove deletion finalizer from %s: %w", df.GetName(), err)
+	}
+	return nil
+}
+
+func (r *AtlasDataFederationReconciler) deleteDataFederationFromAtlas(ctx context.Context, df *mdbv1.AtlasDataFederation, project *mdbv1.AtlasProject, log *zap.SugaredLogger) error {
+	log.Infof("Deleting DataFederation instance: %s from Atlas", df.Spec.Name)
+	connection, err := atlas.ReadConnection(log, r.Client, r.GlobalAPISecret, project.ConnectionSecretObjectKey())
+	if err != nil {
+		return err
+	}
+
+	atlasClient, err := atlas.Client(r.AtlasDomain, connection, log)
+	if err != nil {
+		return fmt.Errorf("can not build Atlas client: %w", err)
+	}
+
+	dfClient := NewClient(atlasClient, r.AtlasDomain)
+
+	_, err = dfClient.Delete(ctx, project.ID(), df.Spec.Name)
+
+	var apiError *mongodbatlas.ErrorResponse
+	if errors.As(err, &apiError) && apiError.Error() == "DATA_LAKE_TENANT_NOT_FOUND_FOR_NAME" {
+		log.Info("DataFederation doesn't exist or is already deleted")
+		return nil
+	}
+
+	if err != nil {
+		log.Errorw("Can not delete Atlas data federation", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+func (r *AtlasDataFederationReconciler) readProjectResource(ctx context.Context, dataFederation *mdbv1.AtlasDataFederation, project *mdbv1.AtlasProject) workflow.Result {
+	if err := r.Client.Get(ctx, dataFederation.AtlasProjectObjectKey(), project); err != nil {
+		return workflow.Terminate(workflow.Internal, err.Error())
+	}
+	return workflow.OK()
+}
+
+func (r *AtlasDataFederationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := controller.New("AtlasDataFederation", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource DataFederation & handle delete separately (can't be done now with connection secrets)
+	err = c.Watch(&source.Kind{Type: &mdbv1.AtlasDataFederation{}}, &watch.EventHandlerWithDelete{Controller: r}, r.GlobalPredicates...)
+	if err != nil {
+		return err
+	}
+	err = c.Watch(&source.Kind{Type: &mdbv1.AtlasDataFederation{}}, &handler.EnqueueRequestForObject{}, r.GlobalPredicates...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete implements a handler for the Delete event. (can't be done now with connection secrets)
+func (r *AtlasDataFederationReconciler) Delete(e event.DeleteEvent) error {
```

Comment:
Never understood why we need this. Can't a`deleteConnectionSecrets()` be called just before deleting the data federation (line 148)?

Reply:
This is due to this: "EventHandlerWithDelete is an extension of EnqueueRequestForObject that will _not_ trigger a reconciliation for a Delete event. Instead, it will call an external controller's Delete() method and pass the event argument unchanged."